mmtk/src/gc_trigger.rs

```diff
@@ -0,0 +1,270 @@
+            },
+            Ordering::Relaxed,
+        );
+        trace!("On GC start: alloc'd = {} = reserved now {} pages - reserved after last gc {} pages = {} pages, prev_sweep = {}", self.actual_allocd.load(Ordering::Relaxed), reserved_pages_now, reserved_pages_in_last_gc, reserved_pages_now.saturating_sub(reserved_pages_in_last_gc), self.prev_sweep_full.load(Ordering::Relaxed));
+    }
+    fn on_gc_end(&self, mmtk: &'static MMTK<JuliaVM>) {
+        use crate::mmtk::vm::ActivePlan;
+        let n_mutators = crate::active_plan::VMActivePlan::number_of_mutators();
+
+        let reserved_pages_before_gc = self.last_recorded_reserved_pages.load(Ordering::Relaxed);
+        let reserved_pages_now = mmtk.get_plan().get_reserved_pages();
+        let freed = conversions::pages_to_bytes(
+            reserved_pages_before_gc.saturating_sub(reserved_pages_now),
+        );
+        self.last_recorded_reserved_pages
+            .store(reserved_pages_now, Ordering::Relaxed);
+        trace!("On GC end: freed = {} = reserved before GC {} pages - reserved now {} pages = {} pages", freed, reserved_pages_before_gc, reserved_pages_now, reserved_pages_before_gc.saturating_sub(reserved_pages_now));
+
+        // ported from gc.c -- before sweeping in the original code.
+        // ignore large frontier (large frontier means the bytes of pointers reachable from the remset is larger than the default collect interval)
+        let gc_auto = !mmtk.is_user_triggered_collection();
+        let not_freed_enough = gc_auto
+            && (freed as f64) < (self.actual_allocd.load(Ordering::Relaxed) as f64 * 0.7f64);
+        let mut sweep_full = false;
+        if gc_auto {
+            if not_freed_enough {
+                trace!(
+                    "Not freed enough: double the interval {:?} -> {:?}",
+                    self.interval.load(Ordering::Relaxed),
+                    self.interval.load(Ordering::Relaxed) * 2
+                );
+                self.interval
+                    .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |interval| {
+                        Some(interval * 2)
+                    })
+                    .unwrap();
+            }
+
+            // on a big memory machine, increase max_collect_interval to totalmem / nthreads / 2
+            let mut maxmem = self.total_mem.load(Ordering::Relaxed) / n_mutators / 2;
+            if maxmem < MAX_COLLECT_INTERVAL {
+                maxmem = MAX_COLLECT_INTERVAL;
+            }
+            if self.interval.load(Ordering::Relaxed) > maxmem {
+                sweep_full = true;
+                trace!(
+                    "Force full heap. Clamp interval back to max mem ({} > {}).",
+                    self.interval.load(Ordering::Relaxed),
+                    maxmem
+                );
+                self.interval.store(maxmem, Ordering::Relaxed);
+            }
+        }
+
+        let live_bytes = conversions::pages_to_bytes(reserved_pages_now);
```

Comment:
Could we use the `count_live_bytes_in_gc` feature and have a fairer comparison with stock Julia. Not that we want to do that necessarily, just wondering if it's as simple as that. I'm also wondering how much their heuristics are dependent on the live bytes instead of live pages.

Reply:
Basically the heuristics use these stats: live bytes/pages, allocated bytes/pages, and freed bytes/pages. We can replace live pages with live bytes in the implementation, but we do not have allocated bytes and freed bytes. I don't think it is a good idea to use live bytes with other page-granularity stats.

Reply:
Also I listed the difference between this implementation and the stock Julia's heuristics here: https://github.com/mmtk/mmtk-julia/blob/e5e1db6af642b8030ad3bb31d57d3459c2b03631/mmtk/src/gc_trigger.rs#L18-L30. They are different in many ways.